swift/extractor/SwiftOutputRewrite.cpp

```diff
@@ -0,0 +1,329 @@
+#include "SwiftOutputRewrite.h"
+#include "swift/extractor/SwiftExtractorConfiguration.h"
+
+#include <llvm/ADT/SmallString.h>
+#include <llvm/Support/FileSystem.h>
+#include <llvm/Support/Path.h>
+#include <swift/Basic/OutputFileMap.h>
+#include <swift/Basic/Platform.h>
+#include <unistd.h>
+#include <unordered_set>
+#include <optional>
+#include <iostream>
+
+// Creates a copy of the output file map and updates remapping table in place
+// It does not change the original map file as it is depended upon by the original compiler
+// Returns path to the newly created output file map on success, or None in a case of failure
+static std::optional<std::string> rewriteOutputFileMap(
+    const codeql::SwiftExtractorConfiguration& config,
+    const std::string& outputFileMapPath,
+    const std::vector<std::string>& inputs,
+    std::unordered_map<std::string, std::string>& remapping) {
+  auto newPath = config.tempArtifactDir + '/' + outputFileMapPath;
+
+  // TODO: do not assume absolute path for the second parameter
+  auto outputMapOrError = swift::OutputFileMap::loadFromPath(outputFileMapPath, "");
+  if (!outputMapOrError) {
+    return std::nullopt;
+  }
+  auto oldOutputMap = outputMapOrError.get();
+  swift::OutputFileMap newOutputMap;
+  std::vector<llvm::StringRef> keys;
+  for (auto& key : inputs) {
+    auto oldMap = oldOutputMap.getOutputMapForInput(key);
+    if (!oldMap) {
+      continue;
+    }
+    keys.push_back(key);
+    auto& newMap = newOutputMap.getOrCreateOutputMapForInput(key);
+    newMap.copyFrom(*oldMap);
+    for (auto& entry : newMap) {
+      auto oldPath = entry.getSecond();
+      auto newPath = config.tempArtifactDir + '/' + oldPath;
+      entry.getSecond() = newPath;
+      remapping[oldPath] = newPath;
```

Comment:
maybe not a big deal for performance, but in general you should use the shiny new C++17 `insert_or_assign`, as it will avoid one default construction and one assignment in case the key is not in the map. And I would move both variables as we don't need it any more, and we then avoid lots of string copies.
```suggestion
      remapping.insert_or_assign(std::move(oldPath), std::move(newPath));
```